src/main/java/com/google/sps/utility/AuthenticationUtility.java

```diff
@@ -0,0 +1,112 @@
+   * @return true if a valid user token is present, false if user needs to reauthenticate
+   * @throws GeneralSecurityException If something goes wrong with the verifier
+   * @throws IOException If something goes wrong with the verifier
+   * @throws IllegalArgumentException if clientID not present
+   */
+  public static boolean verifyUserToken(HttpServletRequest request)
+      throws GeneralSecurityException, IOException, IllegalArgumentException {
+
+    Cookie clientIdCookie = getCookie(request, "clientId");
+    if (clientIdCookie == null) {
+      throw new IllegalArgumentException("Client ID not found!");
+    }
+    String clientId = clientIdCookie.getValue();
+
+    // If idToken not present, user needs to reauthenticate.
+    Cookie idTokenCookie = getCookie(request, "idToken");
+    if (idTokenCookie == null) {
+      return false;
+    }
+
+    String idTokenString = idTokenCookie.getValue();
+
+    // Build a verifier used to ensure the passed user ID is legitimate
+    JacksonFactory jacksonFactory = new JacksonFactory();
+    HttpTransport transport = new NetHttpTransport();
+    GoogleIdTokenVerifier verifier =
+        new GoogleIdTokenVerifier.Builder(transport, jacksonFactory)
+            .setAudience(Collections.singletonList(clientId))
+            .build();
+
+    // If the idToken is not null, the identity is verified and vice versa
+    GoogleIdToken idToken;
+    idToken = verifier.verify(idTokenString);
+
+    return idToken != null;
+  }
+
+  /**
+   * Helper function to create an authorization header for an outgoing HTTP request Does NOT check
+   * if the access token provides correct permissions for the request Does NOT check if the
+   * userToken was valid. Use verifyUserToken to verify userId validity
+   *
+   * @param request contains cookies for a userToken and accessToken
+   * @return Value of the "Authorization" header. Null if authentication is invald or accessToken
+   *     was not found
+   */
+  public static String generateAuthorizationHeader(HttpServletRequest request) {
+    // If accessToken cannot be found, return null
+    Cookie authCookie = getCookie(request, "accessToken");
+    if (authCookie == null) {
+      return null;
+    }
+
+    // Otherwise, return the accessToken as a Bearer token for the Authorization Header
+    return "Bearer " + authCookie.getValue();
+  }
+
+  /**
+   * Helper method to get a cookie from an HttpServletRequest
+   *
+   * @param request HttpServletRequest that contains desired cookie
+   * @param cookieName name of desired cookie. Case sensitive
+   * @return Cookie if found, null if not found or if duplicates present
+   */
+  public static Cookie getCookie(HttpServletRequest request, String cookieName) {
+    List<Cookie> cookies = Arrays.asList(request.getCookies());
+    cookies =
+        cookies.stream()
+            .filter((Cookie c) -> c.getName().equals(cookieName))
+            .collect(Collectors.toList());
```

Comment:
For future reference: this is totally fine, but sometimes it is more performant to postpone 'collecting' the stream until you actually need to access the items to avoid potentially unnecessary processing.

Reply:
Huh interesting - I changed up my code a bit. Let me know if what I did is more performant, I'm curious